src/projections.jl

```diff
@@ -0,0 +1,74 @@
+"""
+    project(M::Manifold, p)
+
+Project point `p` from the ambient space of the [`Manifold`](@ref) `M` to `M`.
+This method is only available for manifolds where implicitly an embedding or ambient space
+is given. Additionally, the projection includes changing data representation, if applicable,
+i.e. if the points on `M` are not represented in the same array data, the data is changed
+accordingly.
+
+See also: [`EmbeddedManifold`](@ref), [`embed`](@ref embed(M::Manifold, p))
+"""
+function project(M::Manifold, p)
+    q = allocate_result(M, project, p)
+    project!(M, q, p)
+    return q
+end
+
+"""
+    project!(M::Manifold, q, p)
+
+Project point `p` from the ambient space onto the [`Manifold`](@ref) `M`. The result is
+storedin `q`.
+This method is only available for manifolds where implicitly an embedding or ambient space
+is given. Additionally, the projection includes changing data representation, if applicable,
+i.e. if the points on `M` are not represented in the same array data, the data is changed
+accordingly.
+
+See also: [`EmbeddedManifold`](@ref), [`embed!`](@ref embed!(M::Manifold, q, p))
+"""
+function project!(M::Manifold, q, p)
+    return error(manifold_function_not_implemented_message(M, project!, q, p))
+end
+
+"""
+    project(M::Manifold, p, X)
+
+Project ambient space representation of a vector `X` to a tangent vector at point `p` on
+the [`Manifold`](@ref) `M`.
+This method is only available for manifolds where implicitly an embedding or ambient space
+is given.
+Additionally, `project` includes changing data representation, if applicable, i.e.
+if the tangents on `M` are not represented in the same way as points on the embedding,
+the representation is changed accordingly. This is the case for example for Lie groups,
+when tangent vectors are represented in the Lie algebra. after projection the change to the
+Lie algebra is perfomed, too.
+
+See also: [`EmbeddedManifold`](@ref), [`embed`](@ref embed(M::Manifold, p, X))
```

Comment:
Is projecting a vector using this and projecting a point from the tangent space manifold using like in https://github.com/JuliaNLSolvers/Optim.jl/issues/920#issuecomment-841662515 the same thing or not? Intuitively projecting to the tangent space may not entail representation change.

Reply:
I don‘t get your second point; If you implement `TangentSpaceAtPoint` to use the same representation as the tangent space at `p` in the usual sense, then the projection of the first should also change representation...?

Reply:
Intuitively, to be consistent, the rules for projection onto `TangentSpaceAtPoint` should follow the rules for projection onto a manifold.

Before I continue, do you think we should have for points `p that are `is_point(M::AbstractEmbedded, p)==true`  the equality `project(M::AbstractEmbedded, p) == p`? It's very easy to assume that's true but representation change would violate this. In any case, I think this should be documented.

Reply:
Also, at the moment `embed` for `Rotations` doesn't revert the transformation but it seems like it should.

Reply:
Yes, but that's more like something over on Manifolds ;)

Reply:
Fixing that problem is something for Manifolds.jl, right, and I think we still need to discuss some points about projections and embeddings but maybe let's leave it to a different PR.